### PR TITLE
Adds evaluation on challenging WxBS and EVD datasets

### PR DESCRIPTION
**Pull request by ducha-aiki**
Closes https://github.com/cvg/glue-factory/issues/35

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -188,6 +188,95 @@ AP_lines: 69.22
...
+```bash
+python -m gluefactory.eval.WxBS --conf gluefactory/configs/superpoint+lightglue-official.yaml
+```
+You should expect the following results
+```
+{'epi_error@10px': 0.6141352941176471,
+ 'epi_error@1px': 0.2968,
+ 'epi_error@20px': 0.6937882352941176,
+ 'epi_error@5px': 0.5143617647058826,
+ 'epi_error_mAA': 0.5297713235294118,
```
Comment (Phil26AT) [inline README.md:257]: When I run this I get different results, close to what is reported in the table below

File: gluefactory/eval/utils.py [hunk 2]
```diff
@@ -69,6 +69,44 @@ def eval_matches_epipolar(data: dict, pred: dict) -> dict:
...
+    success = est["success"] and (len(est["inliers"]) > 0)
+    if not success:
+        results["epi_error"] = [1e6 for i in range(len(data['pts_0to1']))]
+        results["ransac_inl"] = 0
+        results["ransac_inl%"] = 0
+    else:
+        M = est["M_0to1"]
+        inl = est["inliers"].numpy()
+        n_epi_err = sym_epipolar_distance(data['pts_0to1'][:,:2].double(), data['pts_0to1'][:,2:].double(), M.double(), squared=False).detach().cpu().numpy()
+        results["epi_error"] = n_epi_err
```
Comment (Phil26AT) [inline gluefactory/eval/utils.py:99]: results should be a float per entry (here it is a list). I suggest computing the AUC over the epipolar errors here, create one entry per threshold, and then just average them over all data points in the summary. This way it is compatible with the current setup.

File: gluefactory/eval/eval_pipeline.py [hunk 3]
```diff
@@ -25,8 +25,15 @@ def save_eval(dir, summaries, figures, results):
...
-            hfile.create_dataset(k, data=arr)
+                if not isinstance(v[0], str):
+                    arr = np.array([x.astype(np.float64) for x in v])
+                    dt = h5py.special_dtype(vlen=np.float64)
+                    hfile.create_dataset(k, data=arr, dtype=dt)
+                else:
+                    arr = arr.astype("object")
+                    hfile.create_dataset(k, data=arr)
+            else:
+                hfile.create_dataset(k, data=arr)
```
Comment (Phil26AT) [inline gluefactory/eval/eval_pipeline.py:36]: With the fixes proposed in utils this should not be required.

File: gluefactory/eval/utils.py [hunk 4]
```diff
@@ -224,6 +262,36 @@ def eval_poses(pose_results, auc_ths, key, unit="°"):
...
 
 
+def eval_fundamental_matrices(fm_results, auc_ths, key, unit="°"):
+    pose_aucs = {}
+    best_th = -1
+    for th, results_i in fm_results.items():
+        pair_mean = []
+        for pair_results in results_i[key]:
+            pair_mean.append(AUCMetric(auc_ths, pair_results).compute())
+        pose_aucs[th] = np.array(pair_mean).mean(axis=0)
```
Comment (Phil26AT) [inline gluefactory/eval/utils.py:272]: It would be better to compute the AUC over epipolar errors in `eval_matches_epipolar_via_gt_points`, and then just average the respective value here. The only thing this function needs to compute is the mAA, where you can just average all entries in summary that start with `epi_error@X`.

**Review by Phil26AT — CHANGES_REQUESTED** [review]
Thank you for the great contribution, I believe the evals on your challenging datasets are a great addition to gluefactory! 

I marked some improvements in the code. When running the eval on both EVD / WxBS, I get many warnings "libpng warning: iCCP: known incorrect sRGB profile", not sure how this could be adressed. Also, WxBS failed for me, related to `epi_error`. The proposed changes should fix these problems.

Please also run the formatter once such that the CI passes:

```
python -m black .
python -m isort .
```

File: README.md (hunk 1 above)
Comment (ducha-aiki) [reply README.md:257]: I will update this

File: gluefactory/eval/utils.py (hunk 2 above)
Comment (ducha-aiki) [reply gluefactory/eval/utils.py:99]: @Phil26AT unfortunately, that would result in a different number, because different images have very different number of the labeled keypoint correspondences. 
The way it is done now is in line with the paper, whereas one entry for threshold is not, I am afraid.

**Comment by ducha-aiki** [comment]
Thank you for review, I will make changes shortly, where possible. 

Regarding the  "libpng warning: iCCP: known incorrect sRGB profile" - that is an issue with the images themselves, I will check if anybody still have an access to that server to update them.
